cmd/vic-machine/ops-user/rbac_test.go

```diff
@@ -0,0 +1,177 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package opsuser
+
+import (
+	"context"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/vmware/govmomi/simulator"
+	"github.com/vmware/govmomi/vim25/types"
+	"github.com/vmware/vic/pkg/vsphere/session"
+	"github.com/vmware/vic/pkg/vsphere/test/env"
+)
+
+func TestRolesSimulatorVPX(t *testing.T) {
+	ctx := context.Background()
+	m := simulator.VPX()
+
+	defer m.Remove()
+
+	err := m.Create()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	s := m.Service.NewServer()
+	defer s.Close()
+
+	config := &session.Config{
+		Service:   s.URL.String(),
+		Insecure:  true,
+		Keepalive: time.Duration(5) * time.Minute,
+	}
+
+	session, err := session.NewSession(config).Connect(ctx)
+
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	mgr := NewAuthzManager(ctx, session, "ops-user")
+
+	doTestRoles(ctx, t, mgr)
+}
+
+func TestRolesVCenter(t *testing.T) {
+	ctx := context.Background()
+
+	config := &session.Config{
+		Service:   env.URL(t),
+		Insecure:  true,
+		Keepalive: time.Duration(5) * time.Minute,
+	}
+
+	session, err := session.NewSession(config).Connect(ctx)
+
+	if err != nil {
+		t.SkipNow()
+	}
+
+	mgr := NewAuthzManager(ctx, session, "ops-user")
+
+	doTestRoles(ctx, t, mgr)
+}
+
+func initRoles(ctx context.Context, t *testing.T, am *AuthzManager,
+	tRoles []types.AuthorizationRole) int {
+	cleanup(ctx, t, am, false)
+
+	count, err := am.createOrRepairRoles(ctx, tRoles)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	return count
+}
+
+func doTestRoles(ctx context.Context, t *testing.T, am *AuthzManager) {
+	targetRoles := getTargetRoles()
+	var roleCount = len(targetRoles)
+	count := initRoles(ctx, t, am, targetRoles)
+
+	defer cleanup(ctx, t, am, true)
+
+	if count != roleCount {
+		t.Fatalf("Incorrect number of roles: expcted %d, actual %d", roleCount, count)
+	}
+
+	// Test correct role validation, it should return 0
+	roleCount = 0
+	count, err := am.createOrRepairRoles(ctx, targetRoles)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if count != roleCount {
+		t.Fatalf("Incorrect number of roles: expcted %d, actual %d", roleCount, count)
+	}
+
+	// Remove two Privileges from two roles
+	roles, err := am.getRoleList(ctx)
+	fmt.Println(err)
+	fmt.Println(roles)
+
+	targetRoleName1 := rolePrefix + "datastore"
+	targetRoleName2 := rolePrefix + "endpoint"
+
+	for _, role := range roles {
+		if role.Name == targetRoleName1 {
+			removePrivilege(&role, "Datastore.DeleteFile")
+			am.authzManager.UpdateRole(ctx, role.RoleId, role.Name, role.Privilege)
+		}
+		if role.Name == targetRoleName2 {
+			removePrivilege(&role, "VirtualMachine.Config.AddNewDisk")
+			am.authzManager.UpdateRole(ctx, role.RoleId, role.Name, role.Privilege)
+		}
+	}
+
+	// Test
+	roleCount = 2
+	count, err = am.createOrRepairRoles(ctx, targetRoles)
```

Comment:
There are `testify/assert` (continue but fail test) and `testify/require` (fail immediately and stop) packages that simplify this conditional logic and allow for embedding textual commentary in the test output. It makes the tests much more declarative.

Reply:
I left one explicit check. It is:

if checkCount && count != roleCount {
   // fail
}

I tried require.True() but the code becomes less readable in my opinion.